// ngx-scrollbar/esm5/lib/scrollbar/track/track.js
import { __assign } from "tslib";
import { coerceNumberProperty } from '@angular/cdk/coercion';
import { EMPTY, fromEvent, merge, of } from 'rxjs';
import { map, pluck, switchMap, tap } from 'rxjs/operators';
import { preventSelection, enableSelection, stopPropagation } from '../common';
var TrackAdapter = /** @class */ (function () {
    function TrackAdapter(cmp, trackElement, document) {
        this.cmp = cmp;
        this.trackElement = trackElement;
        this.document = document;
    }
    Object.defineProperty(TrackAdapter.prototype, "clicked", {
        // Stream that emits when the track element is clicked
        get: function () {
            var mouseDown = fromEvent(this.trackElement, 'mousedown', { passive: true }).pipe(stopPropagation(), preventSelection(this.document));
            var mouseup = fromEvent(this.document, 'mouseup', { passive: true }).pipe(stopPropagation(), enableSelection(this.document), switchMap(function () { return EMPTY; }));
            return merge(mouseDown, mouseup);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TrackAdapter.prototype, "hovered", {
        // Stream that emits when the track element is hovered
        get: function () {
            var mouseEnter = fromEvent(this.trackElement, 'mouseenter', { passive: true }).pipe(stopPropagation(), map(function () { return true; }));
            var mouseLeave = fromEvent(this.trackElement, 'mouseleave', { passive: true }).pipe(stopPropagation(), map(function () { return false; }));
            return merge(mouseEnter, mouseLeave);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TrackAdapter.prototype, "clientRect", {
        // Get track client rect
        get: function () {
            return this.trackElement.getBoundingClientRect();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Stream that emits when scrollbar track is clicked
     */
    TrackAdapter.prototype.onTrackClicked = function (e, thumbSize, scrollSize) {
        var _this = this;
        return of(e).pipe(pluck(this.pageProperty), 
        // Calculate scrollTo position
        map(function (pageOffset) {
            var clickOffset = pageOffset - _this.offset;
            var offset = clickOffset - (thumbSize / 2);
            var ratio = offset / _this.size;
            return ratio * scrollSize;
        }), 
        // Smooth scroll to position
        tap(function (value) {
            _this.cmp.scrollTo(__assign(__assign({}, _this.mapToScrollToOption(value)), { duration: coerceNumberProperty(_this.cmp.trackClickScrollDuration) }));
        }));
    };
    return TrackAdapter;
}());
export { TrackAdapter };
