// ngx-scrollbar/esm5/lib/scrollbar/scrollbar.component.js
import { __decorate, __extends, __metadata, __param } from "tslib";
import { Component, Inject, NgZone, ChangeDetectionStrategy, ElementRef, ViewChild } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import { Platform } from '@angular/cdk/platform';
import { TrackXDirective, TrackYDirective } from './track/track.directive';
import { ThumbXDirective, ThumbYDirective } from './thumb/thumb.directive';
import { NgScrollbar } from '../ng-scrollbar';
import { Scrollbar } from './scrollbar';
var ScrollbarY = /** @class */ (function (_super) {
    __extends(ScrollbarY, _super);
    function ScrollbarY(cmp, platform, document, zone) {
        var _this = _super.call(this, cmp, platform, document, zone) || this;
        _this.cmp = cmp;
        _this.platform = platform;
        _this.document = document;
        _this.zone = zone;
        return _this;
    }
    Object.defineProperty(ScrollbarY.prototype, "viewportScrollSize", {
        get: function () {
            return this.cmp.viewport.scrollHeight;
        },
        enumerable: true,
        configurable: true
    });
    ScrollbarY.prototype.setHovered = function (value) {
        this.cmp.setHovered({ verticalHovered: value });
    };
    ScrollbarY.prototype.onUpdated = function () {
    };
    ScrollbarY.ctorParameters = function () { return [
        { type: NgScrollbar },
        { type: Platform },
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
        { type: NgZone }
    ]; };
    __decorate([
        ViewChild(TrackYDirective, { static: true }),
        __metadata("design:type", TrackYDirective)
    ], ScrollbarY.prototype, "track", void 0);
    __decorate([
        ViewChild(ThumbYDirective, { static: true }),
        __metadata("design:type", ThumbYDirective)
    ], ScrollbarY.prototype, "thumb", void 0);
    ScrollbarY = __decorate([
        Component({
            selector: 'scrollbar-y',
            host: { '[class.scrollbar-control]': 'true' },
            changeDetection: ChangeDetectionStrategy.OnPush,
            template: "\n    <div scrollbarTrackY class=\"ng-scrollbar-track {{cmp.trackClass}}\">\n      <div scrollbarThumbY [track]=\"track\" class=\"ng-scrollbar-thumb {{cmp.thumbClass}}\"></div>\n    </div>\n  ",
            styles: ["::ng-deep .ng-scrollbar-wrapper>scrollbar-y.scrollbar-control{width:var(--vertical-scrollbar-total-size)}::ng-deep .ng-scrollbar-wrapper>scrollbar-y.scrollbar-control>.ng-scrollbar-track{width:var(--vertical-scrollbar-size);height:calc(100% - var(--scrollbar-padding) * 2)}::ng-deep .ng-scrollbar-wrapper>scrollbar-y.scrollbar-control>.ng-scrollbar-track>.ng-scrollbar-thumb{height:0;width:100%}::ng-deep .ng-scrollbar-wrapper[verticalDragging=true]>scrollbar-y.scrollbar-control .ng-scrollbar-thumb,::ng-deep .ng-scrollbar-wrapper[verticalHovered=true]>scrollbar-y.scrollbar-control .ng-scrollbar-thumb{background-color:var(--scrollbar-thumb-hover-color)}::ng-deep .ng-scrollbar-wrapper[deactivated=false]>scrollbar-y.scrollbar-control{top:0;bottom:0}::ng-deep .ng-scrollbar-wrapper[deactivated=false][dir=ltr]>scrollbar-y.scrollbar-control{right:0;left:unset}::ng-deep .ng-scrollbar-wrapper[deactivated=false][dir=ltr][position=invertAll]>scrollbar-y.scrollbar-control,::ng-deep .ng-scrollbar-wrapper[deactivated=false][dir=ltr][position=invertY]>scrollbar-y.scrollbar-control{left:0;right:unset}::ng-deep .ng-scrollbar-wrapper[deactivated=false][dir=rtl]>scrollbar-y.scrollbar-control{left:0;right:unset}::ng-deep .ng-scrollbar-wrapper[deactivated=false][dir=rtl][position=invertAll]>scrollbar-y.scrollbar-control,::ng-deep .ng-scrollbar-wrapper[deactivated=false][dir=rtl][position=invertY]>scrollbar-y.scrollbar-control{left:unset;right:0}::ng-deep .ng-scrollbar-wrapper[deactivated=false][track=all]>scrollbar-y.scrollbar-control[fit=true]{bottom:var(--scrollbar-total-size);top:0}::ng-deep .ng-scrollbar-wrapper[deactivated=false][track=all][position=invertAll]>scrollbar-y.scrollbar-control[fit=true],::ng-deep .ng-scrollbar-wrapper[deactivated=false][track=all][position=invertX]>scrollbar-y.scrollbar-control[fit=true]{top:var(--scrollbar-total-size);bottom:0}"]
        }),
        __param(2, Inject(DOCUMENT)),
        __metadata("design:paramtypes", [NgScrollbar, Platform, Object, NgZone])
    ], ScrollbarY);
    return ScrollbarY;
}(Scrollbar));
export { ScrollbarY };
var ScrollbarX = /** @class */ (function (_super) {
    __extends(ScrollbarX, _super);
    function ScrollbarX(el, cmp, platform, document, zone) {
        var _this = _super.call(this, cmp, platform, document, zone) || this;
        _this.el = el;
        _this.cmp = cmp;
        _this.platform = platform;
        _this.document = document;
        _this.zone = zone;
        return _this;
    }
    Object.defineProperty(ScrollbarX.prototype, "viewportScrollSize", {
        get: function () {
            return this.cmp.viewport.scrollWidth;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ScrollbarX.prototype, "thickness", {
        get: function () {
            return this.el.nativeElement.clientHeight;
        },
        enumerable: true,
        configurable: true
    });
    ScrollbarX.prototype.setHovered = function (value) {
        this.cmp.setHovered({ horizontalHovered: value });
    };
    ScrollbarX.prototype.onUpdated = function () {
        // Auto-height: Set root component height to content height
        this.cmp.nativeElement.style.height = this.cmp.appearance === 'standard'
            ? this.cmp.viewport.contentHeight + this.thickness + "px"
            : this.cmp.viewport.contentHeight + "px";
    };
    ScrollbarX.ctorParameters = function () { return [
        { type: ElementRef },
        { type: NgScrollbar },
        { type: Platform },
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
        { type: NgZone }
    ]; };
    __decorate([
        ViewChild(TrackXDirective, { static: true }),
        __metadata("design:type", TrackXDirective)
    ], ScrollbarX.prototype, "track", void 0);
    __decorate([
        ViewChild(ThumbXDirective, { static: true }),
        __metadata("design:type", ThumbXDirective)
    ], ScrollbarX.prototype, "thumb", void 0);
    ScrollbarX = __decorate([
        Component({
            selector: 'scrollbar-x',
            host: { '[class.scrollbar-control]': 'true' },
            changeDetection: ChangeDetectionStrategy.OnPush,
            template: "\n    <div scrollbarTrackX class=\"ng-scrollbar-track {{cmp.trackClass}}\">\n      <div scrollbarThumbX [track]=\"track\" class=\"ng-scrollbar-thumb {{cmp.thumbClass}}\"></div>\n    </div>\n  ",
            styles: ["::ng-deep .ng-scrollbar-wrapper>scrollbar-x.scrollbar-control{height:var(--horizontal-scrollbar-total-size)}::ng-deep .ng-scrollbar-wrapper>scrollbar-x.scrollbar-control>.ng-scrollbar-track{height:var(--horizontal-scrollbar-size);width:calc(100% - var(--scrollbar-padding) * 2)}::ng-deep .ng-scrollbar-wrapper>scrollbar-x.scrollbar-control>.ng-scrollbar-track>.ng-scrollbar-thumb{width:0;height:100%}::ng-deep .ng-scrollbar-wrapper[horizontalDragging=true]>scrollbar-x.scrollbar-control .ng-scrollbar-thumb,::ng-deep .ng-scrollbar-wrapper[horizontalHovered=true]>scrollbar-x.scrollbar-control .ng-scrollbar-thumb{background-color:var(--scrollbar-thumb-hover-color)}::ng-deep .ng-scrollbar-wrapper[position=invertAll]>scrollbar-x.scrollbar-control,::ng-deep .ng-scrollbar-wrapper[position=invertX]>scrollbar-x.scrollbar-control{top:0;bottom:unset}::ng-deep .ng-scrollbar-wrapper[deactivated=false]>scrollbar-x.scrollbar-control{left:0;right:0;bottom:0;top:unset}::ng-deep .ng-scrollbar-wrapper[deactivated=false][position=invertAll]>scrollbar-x.scrollbar-control,::ng-deep .ng-scrollbar-wrapper[deactivated=false][position=invertX]>scrollbar-x.scrollbar-control{top:0;bottom:unset}::ng-deep .ng-scrollbar-wrapper[deactivated=false][track=all][dir=ltr]>scrollbar-x.scrollbar-control[fit=true]{right:var(--scrollbar-total-size);left:0}::ng-deep .ng-scrollbar-wrapper[deactivated=false][track=all][dir=ltr][position=invertAll]>scrollbar-x.scrollbar-control[fit=true],::ng-deep .ng-scrollbar-wrapper[deactivated=false][track=all][dir=ltr][position=invertY]>scrollbar-x.scrollbar-control[fit=true]{left:var(--scrollbar-total-size);right:0}::ng-deep .ng-scrollbar-wrapper[deactivated=false][track=all][dir=rtl]>scrollbar-x.scrollbar-control[fit=true]{left:var(--scrollbar-total-size);right:0}::ng-deep .ng-scrollbar-wrapper[deactivated=false][track=all][dir=rtl][position=invertAll]>scrollbar-x.scrollbar-control[fit=true],::ng-deep .ng-scrollbar-wrapper[deactivated=false][track=all][dir=rtl][position=invertY]>scrollbar-x.scrollbar-control[fit=true]{right:var(--scrollbar-total-size);left:0}"]
        }),
        __param(3, Inject(DOCUMENT)),
        __metadata("design:paramtypes", [ElementRef,
            NgScrollbar,
            Platform, Object, NgZone])
    ], ScrollbarX);
    return ScrollbarX;
}(Scrollbar));
export { ScrollbarX };
